book/core-concepts/windows.md

```diff
@@ -0,0 +1,152 @@
+# Windows
+
+Wallaroo supports windowing over aggregations. This allows you to break an infinite stream into manageable chunks and also allows you to see how your inputs evolve over time.
+
+There are two broad categories of windows currently supported: count-based and range-based. Range-based windows can be further divided into tumbling and sliding varieties. We'll look at each of these in turn.
+
+If you are not familiar with aggregations in Wallaroo, it might be helpful to read [this section](/book/core-concepts/aggregations.md) before proceeding.
+
+## Count-based Windows
+
+Count-based windows emit an output every `n` input messages, where `n` is specified via the API. For example (using the Python API):
+
+```python
+    (inputs
+        .to(wallaroo.count_windows(5)
+            .over(MySumAgg))
+        .to_sink(sink_config))
+```
+
+This performs the user-defined `MySumAgg` aggregation for every 5 inputs. 
+
+```
+Messages:
+01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 
+|-------------|--------------|--------------|--------------|
+    window 1      window 2       window 3       window 4
+    sum = 15      sum = 40       sum = 65       sum = 90
+```
+
+## Range-based Windows
+
+Range-based windows are defined based on specific time-based ranges. They fall into two categories, tumbling and sliding. We'll look at tumbling first.
+
+### Tumbling Windows
+
+Tumbling range-based windows are non-overlapping fixed-size time-based windows.
+For example, let's define windows with a range of 3 seconds:
+
+```python
+    (inputs
+        .to(wallaroo.range_windows(wallaroo.seconds(3))
+            .over(MySumAgg))
+        .to_sink(...))
+```
+
+Assuming we start from time 00:00, this would produce the following tumbling windows:
+
+```
+ Time (min:sec):
+|00:00|00:01|00:02|00:03|00:04|00:05|00:06|00:07|00:08|00:09|00:10|00:11
+|-----------------|-----------------|-----------------|-----------------|
+     window 1          window 2          window 3          window 4
+   [00:00,00:03)     [00:03,00:06)     [00:06,00:09)     [00:09,00:12)
+```
+
+Each window above is indicated by an interval such as `[00:00,00:03)`, indicating "the window containing all times equal to or greater than `00:00` and less than `00:03`".
+
+Incoming messages will be placed in a window based on the event time of the message. In the near future, you will be able to derive this event time from each input message that arrives at a Wallaroo source. For now, it is always the ingest time for the input message.
+
+So if we had the following four messages (min:sec:ms):
```

Comment:
Nit: I'm used to the `min:sec.ms` notation.
The `00:00:000` notation made me think of `HH:MM:ss` notation and I think this could be confusing to others as well.

Reply:
addressed